memote/support/thermodynamics.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 Novo Nordisk Foundation Center for Biosustainability,
+# Technical University of Denmark.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Supporting functions for checks requiring the eQuilibrator API."""
+
+from __future__ import absolute_import
+
+from sys import version_info
+
+from six import string_types
+
+
+if version_info[:2] >= (3, 5):
+    from equilibrator_api import Reaction, CompoundMatcher
+    COMPOUND_MATCHER = CompoundMatcher()
+
+
+def smallest_compound_ID(kegg_ann_list):
+    """
+    Return an ascending list filtered to contain only KEGG compound IDs.
+
+    KEGG IDs may map to compounds, drugs and glycans prefixed respectively
+    with "C", "D", and "G" followed by at least 5 digits.
+
+    Parameters
+    ----------
+    kegg_ann_list: list
+        A list of mixed KEGG IDs.
+
+    """
+    only_cIDs = [x for x in kegg_ann_list if "C" in x]
+    only_cIDs.sort(key=lambda x: int(x.lstrip('C')))
+    return only_cIDs
+
+
+def get_equilibrator_rxn_string(rxn):
+    """
+    Return a reaction string with at least a partial mapping to KEGG IDs.
+
+    First see if there is an unambiguous mapping to a single KEGG compound ID
+    provided with the model. If not, check if there is any KEGG compound ID in
+    a list of mappings. KEGG IDs may map to compounds, drugs and glycans. KEGG
+    compound IDs are sorted so we keep the lowest that is there. If none of
+    this works try mapping to KEGG via the CompoundMatcher by the name of the
+    metabolite. If the metabolite cannot be mapped we simply don't replace it
+    in the original reaction string.
+
+    Parameters
+    ----------
+    rxn: cobra.Reaction
+        The metabolic reaction under investigation.
+
+    """
+    kegg_rxn = rxn.reaction
+    for met in rxn.metabolites:
```

Comment:
I would structure this helper function differently. You are calling this function for each reaction `r` and looping through its metabolites `m[r]`. So that means you construct `O(r * m[r])` mappings. I would create a mapping object only from the purely metabolic metabolites and pass that into this function to construct the reaction from. That means you first loop through metabolites `m` after which you loop through reactions `r` and thus  `O(r + m)` complexity. I have also [opened an issue](https://gitlab.com/elad.noor/equilibrator-api/issues/18) to create some kind of adapter between `cobra.Reaction` and equilibrator reactions.

Reply:
I'm a big fan of the adapter idea, but until then I'm adamant about keeping the crutch (as opposed to stalling this).

I'll consider your suggestion, shouldn't be too hard to do.

Reply:
The newest commit should have addressed your concerns here. I'm hence resolving this conversation.